### ISSUE TITLE
[BUG] Gmail does not encode UTF-8 or use A-labels for IDNs in To or RCPT TO

### ISSUE THREAD

**Issue by TaaviE** [rspamd/rspamd#5002]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
* [X] Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
* [X] Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
* [X] Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
* [X] Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
* [X] Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
* [X] Checked that there is not already an experimental package or master branch

**Describe the bug**
Gmail no longer performs any modifications on UTF-8 recipients. This results in a relatively high score for any UTF-8/IDN recipients.

### Steps to Reproduce

Sending a letter from Gmail to `äää@äää.ee` will result in the following symbols: `BROKEN_HEADERS`, `INVALID_RCPT_8BIT`, `TO_NEEDS_ENCODING`, `RCVD_ILLEGAL_CHARS`

**Expected behavior**
In theory such letters are not violating any RFC (especially with `SMTPUTF8` in the mix), the local part can't be encoded, the IDN should be okay as u-label.

Maybe a symbol like `UTF8_IDN_RECIPIENT` would be sufficient to handle these cases, but when an UTF-8 local part is combined with an IDN, it shouldn't be punished. 

### Versions

3.8.4

**Comment by vstakhov** [comment]
If gmail does not encode 8bit characters in the headers it is their problem.

**Comment by TaaviE** [comment]
@vstakhov There's actually no RFC2047 encoding of addresses themselves and in the case of SMTPUTF8 just UTF-8 should be accepted.

**Comment by vstakhov** [comment]
I don't understand how is MIME encoding (meaning message based) related to SMTPUTF (meaning the concrete message transaction based).

**Comment by vstakhov** [comment]
All message headers must be 7bit. Apparently rfc2047 allows to encode any header, including `From`, `To` and even `Received`.

**Comment by vstakhov** [comment]
The only thing that actually allows utf8 in the headers is SMTPUTF rfc. Rspamd does not support it and I have no intention to add such a support (unless it is absolutely necessary due to ubiquotous usage) as from my point of view this standard is designed by brain damaged people. 

Rspamd is UTF8 compatible by design (that was my major goal from the begining), it is just about this particular standard - it is broken, ambiguious and conflicting with the basics of the SMTP design principles (at least in terms of end-to-end principle).

**Comment by TaaviE** [comment]
@vstakhov 

> Apparently rfc2047 allows to encode any header, including From, To and even Received.

No. It's only allowed for the 'comment' or 'phrase' part of those headers, not the structured part of the header (the addresses).

> [5](https://www.rfc-editor.org/rfc/rfc2047#section-5). Use of encoded-words in message headers
> As a replacement for a 'word' entity within a 'phrase', for example,
>    one that precedes an address in a From, To, or Cc header.  The ABNF
>    definition for 'phrase' from [RFC 822](https://www.rfc-editor.org/rfc/rfc822) thus becomes:

In addition:

>   An 'encoded-word' MUST NOT be used in a Received header field.

**Comment by TaaviE** [comment]
The end result is that letters to recipients with UTF-8 local parts will use UTF-8 in headers that can't be encoded RFC-compliant and that seems to be expected behavior. IDNs thankfully have an A-label which allows some fallback (as far as I can see), but with local parts that does not seem to be the case.

**Comment by vstakhov** [comment]
You are right @TaaviE and that's really a bit problematic. It seems we need to support SMTPUTF8 at some point. I don't have much desire to implement that though, however PRs are very wellcome.

**Comment by vstakhov** [comment]
Tracking PR: https://github.com/rspamd/rspamd/pull/5056

**Comment by vstakhov** [comment]
Do you have any .eml samples with UTF8 text inside?

**Comment by vstakhov** [comment]
Can you please send something from this address to my personal gmail `vstakhov`?

**Comment by TaaviE** [comment]
@vstakhov Sent.

**Comment by vstakhov** [comment]
Interesting... In fact, Rspamd supports utf8 in headers almost everywhere. I'm trying to understand where it could be broken.

**Comment by dragoangel** [comment]
We need parse not only headers, but also envelope from to rcpt

**Comment by dragoangel** [comment]
Regarding my message about envelope and utf8 @vstakhov created a fix 🥳: https://github.com/rspamd/rspamd/pull/5210

Testing it now

**Comment by dragoangel** [comment]
From my tests when Postfix correctly sends utf8 encoded data to milter it works great, so rspamd-proxy <-> rspamd-normal works with UTF8.

Also tested setup under:

1. Nginx reverse proxy - there no issues in passing UTF8 values in http headers directly to rspamd-normal
2. Python requests lib can process header like `'веселий-ящик@дім.укр'.encode('utf8')` without issues.

Some less popular or fresh http libs that don't like non-US-ASCII in HTTP header values could be a problem. Having capability to encode UTF8 payload inside of base64 value of `From`, `Rcpt` and `User` headers with prefix like `b64:` that Rspamd will decode and use will be most portable option for some legacy software. For example query with curl will only be possible in this way, mostly sure other scripting tools will also will complain about non-US-ASCII. But as python/rust/go and reverse proxy's works - it's not a big problem really.

**Comment by dragoangel** [comment]
@vstakhov also based on what I see I can propose `enable_mime_utf = true;` to be default in some future release, f.e. in v3.11 or 3.12
